core/src/main/java/org/infinispan/interceptors/distribution/TxDistributionInterceptor.java

```diff
@@ -91,37 +110,32 @@ public Object visitRemoveCommand(InvocationContext ctx, RemoveCommand command) t
       }
    }
 
-   @Start
-   public void start() {
-      isPessimisticCache = cacheConfiguration.transaction().lockingMode() == LockingMode.PESSIMISTIC;
-      useClusteredWriteSkewCheck = !isPessimisticCache &&
-            cacheConfiguration.versioning().enabled() && cacheConfiguration.locking().writeSkewCheck();
-   }
-
    @Override
    public Object visitPutKeyValueCommand(InvocationContext ctx, PutKeyValueCommand command) throws Throwable {
       if (command.hasFlag(Flag.PUT_FOR_EXTERNAL_READ)) {
          return handleNonTxWriteCommand(ctx, command);
       }
 
-      SingleKeyRecipientGenerator skrg = new SingleKeyRecipientGenerator(command.getKey());
-      Object returnValue = handleTxWriteCommand(ctx, command, skrg, command.hasFlag(Flag.PUT_FOR_STATE_TRANSFER));
-      if (ctx.isOriginLocal()) {
-         // If the state transfer interceptor has to retry the command, it should ignore the previous value.
-         command.setValueMatcher(command.isSuccessful() ? ValueMatcher.MATCH_ALWAYS : ValueMatcher.MATCH_NEVER);
+      try {
+         return handleTxWriteCommand(ctx, command, new SingleKeyRecipientGenerator(command.getKey()),
+                                     command.hasFlag(Flag.PUT_FOR_STATE_TRANSFER));
```

Comment:
I think you could have removed the `RecipientGenerator` completely and pass the key directly.